File: Metadata-Handling/app/src/main/java/com/theoplayer/sample/playback/metadata/PlayerActivity.java
package com.theoplayer.sample.playback.metadata;

import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.text.Layout;
import android.text.SpannableString;
import android.text.style.AlignmentSpan;
import android.util.Base64;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintSet;
import androidx.databinding.DataBindingUtil;

import com.theoplayer.android.api.event.player.PlayerEventTypes;
import com.theoplayer.android.api.event.track.texttrack.TextTrackEventTypes;
import com.theoplayer.android.api.event.track.texttrack.list.TextTrackListEventTypes;
import com.theoplayer.android.api.player.Player;
import com.theoplayer.android.api.player.track.texttrack.TextTrackType;
import com.theoplayer.android.api.player.track.texttrack.cue.DateRangeCue;
import com.theoplayer.android.api.source.SourceDescription;
import com.theoplayer.android.api.source.TypedSource;
import com.theoplayer.sample.playback.metadata.databinding.ActivityPlayerBinding;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.util.Iterator;

import static android.util.Base64.NO_WRAP;

public class PlayerActivity extends AppCompatActivity {

    private static final String TAG = PlayerActivity.class.getSimpleName();

    private static final String PLAYER_PARAM__METADATA_ID = "METADATA_ID";

    private ActivityPlayerBinding viewBinding;
    private Player theoPlayer;

    public static void play(Context context, int metadataId) {
        Intent playIntent = new Intent(context, PlayerActivity.class);
        playIntent.putExtra(PLAYER_PARAM__METADATA_ID, metadataId);
        context.startActivity(playIntent);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Inflating view and obtaining an instance of the binding class.
        viewBinding = DataBindingUtil.setContentView(this, R.layout.activity_player);

        // Gathering THEO objects references.
        theoPlayer = viewBinding.theoPlayerView.getPlayer();

        // Configuring action bar.
        setSupportActionBar(viewBinding.toolbarLayout.toolbar);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        // Configuring THEOplayer playback with parameters for appropriate metadata handling.
        switch (getIntent().getIntExtra(PLAYER_PARAM__METADATA_ID, 0)) {
            case R.string.hlsWithID3MetadataName:
                handleHlsWithID3Metadata();
                break;
            case R.string.hlsWithProgramDateTimeMetadataName:
                handleHlsWithProgramDateTimeMetadata();
                break;
            case R.string.hlsWithDateRangeMetadataName:
                handleHlsWithDateRangeMetadata();
                break;
            case R.string.dashWithEmsgMetadataName:
                handleDashWithEmsgMetadata();
                break;
            case R.string.dashWithEventStreamMetadataName:
                handleDashWithEventStreamMetadata();
                break;
            default:
                SpannableString toastMessage = SpannableString.valueOf(this.getString(R.string.missingMetadataConfiguration));
                toastMessage.setSpan(new AlignmentSpan.Standard(Layout.Alignment.ALIGN_CENTER), 0, toastMessage.length(), 0);
                Toast.makeText(this, toastMessage, Toast.LENGTH_LONG).show();
        }
    }

    /**
     * Demonstrates THEOplayer configuration that allows to handle ID3 metadata from HLS stream.
     */
    private void handleHlsWithID3Metadata() {
        viewBinding.headerTextView.setText(getString(R.string.hlsWithID3MetadataHeader));

        // Configuring THEOplayer with appropriate stream source.
        configureTHEOplayer(
                TypedSource.Builder.typedSource(getString(R.string.hlsWithID3MetadataSourceUrl))
        );

        // Listening to 'addtrack' events to find text track of type 'id3'.
        theoPlayer.getTextTracks().addEventListener(TextTrackListEventTypes.ADDTRACK, event -> {

            if (event.getTrack().getType() == TextTrackType.ID3) {
                Log.i(TAG, "Event: ADDTRACK, trackType=" + event.getTrack().getType());

                // Listening to 'exitcue' event. For ID3 metadata exited cue means "current cue".
                event.getTrack().addEventListener(TextTrackEventTypes.EXITCUE, cueEvent -> {
                    Log.i(TAG, "Event: EXITCUE, cue=" + cueEvent.getCue());

                    // Decoding ID3 metadata. In this example, the data received
                    // is in the form: '{"content":{"id":"TXXX","description":"","text":"..."}}}'.
                    JSONObject cueContent = cueEvent.getCue().getContent();
                    try {
                        appendMetadata(cueContent.getJSONObject("content").getString("text"));
                    } catch (JSONException exception) {
                        appendMetadata(cueContent.toString());
                    }
                });
            }

        });
    }

    /**
     * Demonstrates THEOplayer configuration that allows to handle EXT-X-PROGRAM-DATE-TIME metadata from HLS stream.
     */
    private void handleHlsWithProgramDateTimeMetadata() {
        viewBinding.headerTextView.setText(getString(R.string.hlsWithProgramDateTimeMetadataHeader));

        // Configuring THEOplayer with appropriate stream source.
        configureTHEOplayer(
                TypedSource.Builder.typedSource(getString(R.string.hlsWithProgramDateTimeMetadataSourceUrl))
        );

        // Listening to 'timeupdate' events that are triggered every time EXT-X-PROGRAM-DATE-TIME
        // is updated.
        theoPlayer.addEventListener(PlayerEventTypes.TIMEUPDATE, event -> {
            Log.i(TAG, "Event: TIMEUPDATE, currentTime=" + event.getCurrentTime());

            // Once we know that EXT-X-PROGRAM-DATE-TIME was updated we have to request for its value.
            theoPlayer.requestCurrentProgramDateTime(date -> {
                appendMetadata(date == null ? "" : date.toString());
            });

        });
    }

    /**
     * Demonstrates THEOplayer configuration that allows to handle EXT-X-DATERANGE metadata from HLS stream.
     */
    private void handleHlsWithDateRangeMetadata() {
        viewBinding.headerTextView.setText(getString(R.string.hlsWithDateRangeMetadataHeader));

        // Configuring THEOplayer with appropriate stream source. Note that logic that exposes date
        // ranges parsed from HLS manifest needs to be enabled.
        configureTHEOplayer(
                TypedSource.Builder
                        .typedSource(getString(R.string.hlsWithDateRangeMetadataSourceUrl))
                        .hlsDateRange(true)
        );

        // Listening to 'addtrack' events to find text track of type 'daterange'.
        theoPlayer.getTextTracks().addEventListener(TextTrackListEventTypes.ADDTRACK, event -> {

            if (event.getTrack().getType() == TextTrackType.DATERANGE) {
                Log.i(TAG, "Event: ADDTRACK, trackType=" + event.getTrack().getType());

                // Listening to 'addcue' event to get parsed date range.
                event.getTrack().addEventListener(TextTrackEventTypes.ADDCUE, cueEvent -> {
                    DateRangeCue cue = (DateRangeCue) cueEvent.getCue();

                    Log.i(TAG, "Event: ADDCUE, cue=" + cue);

                    // Decoding date range metadata. For demo purposes we are displaying
                    // content as it is encoding byte arrays with base64.
                    appendMetadata("StartDate: " + cue.getStartDate() +
                            "\nEndDate: " + cue.getEndDate() +
                            "\nDuration: " + cue.getDuration() +
                            "\nScte35Cmd: " + (cue.getScte35Cmd() != null ? Base64.encodeToString(cue.getScte35Cmd(), NO_WRAP) : "N/A") +
                            "\nScte35In: " + (cue.getScte35In() != null ? Base64.encodeToString(cue.getScte35In(), NO_WRAP) : "N/A") +
                            "\nScte35Out: " + (cue.getScte35Out() != null ? Base64.encodeToString(cue.getScte35Out(), NO_WRAP) : "N/A"));
                });
            }

        });
    }

    /**
     * Demonstrates THEOplayer configuration that allows to handle EMSG metadata from DASH stream.
     */
    private void handleDashWithEmsgMetadata() {
        viewBinding.headerTextView.setText(getString(R.string.dashWithEmsgMetadataHeader));

        // Configuring THEOplayer with appropriate stream source.
        configureTHEOplayer(
                TypedSource.Builder.typedSource(getString(R.string.dashWithEmsgMetadataSourceUrl))
        );

        // Listening to 'addtrack' events to find text track of type 'emsg'.
        theoPlayer.getTextTracks().addEventListener(TextTrackListEventTypes.ADDTRACK, event -> {

            if (event.getTrack().getType() == TextTrackType.EMSG) {
                Log.i(TAG, "Event: ADDTRACK, trackType=" + event.getTrack().getType());

                // Listening to 'addcue' event to read EMSG metadata.
                event.getTrack().addEventListener(TextTrackEventTypes.ADDCUE, cueEvent -> {
                    Log.i(TAG, "Event: ADDCUE, cue=" + cueEvent.getCue());

                    // Decoding EMSG metadata. In this example, the data received
                    // is in the form: '{"content":{"0":73,"1":68,"2":51,"3":4,"4":0,"5":32,...}'.
                    JSONObject cueContent = cueEvent.getCue().getContent();
                    try {
                        ByteArrayOutputStream byteContent = new ByteArrayOutputStream();
                        JSONObject jsonContent = cueContent.getJSONObject("content");
                        Iterator<String> jsonContentKeys = jsonContent.keys();
                        while (jsonContentKeys.hasNext()) {
                            byteContent.write(jsonContent.getInt(jsonContentKeys.next()));
                        }
                        appendMetadata(new String(byteContent.toByteArray()));
                    } catch (JSONException e) {
                        appendMetadata(cueContent.toString());
                    }
                });
            }

        });
    }

    /**
     * Demonstrates THEOplayer configuration that allows to handle EventStream metadata from DASH stream.
     */
    private void handleDashWithEventStreamMetadata() {
        viewBinding.headerTextView.setText(getString(R.string.dashWithEventStreamMetadataHeader));

        // Configuring THEOplayer with appropriate stream source.
        configureTHEOplayer(
                TypedSource.Builder.typedSource(getString(R.string.dashWithEventStreamMetadataSourceUrl))
        );

        // Listening to 'addtrack' events to find text track of type 'eventstream'.
        theoPlayer.getTextTracks().addEventListener(TextTrackListEventTypes.ADDTRACK, event -> {

            if (event.getTrack().getType() == TextTrackType.EVENTSTREAM) {
                Log.i(TAG, "Event: ADDTRACK, trackType=" + event.getTrack().getType());

                // Listening to 'addcue' event to read EventStream metadata.
                event.getTrack().addEventListener(TextTrackEventTypes.ADDCUE, cueEvent -> {
                    Log.i(TAG, "Event: ADDCUE, cue=" + cueEvent.getCue());

                    // For demo purposes we are displaying whole content as it is.
                    appendMetadata(cueEvent.getCue().getContent().toString());
                });
            }
        });
    }

    private void configureTHEOplayer(TypedSource.Builder typedSource) {
        // Creating a SourceDescription builder that contains the settings to be applied as a new
        // THEOplayer source.
        SourceDescription.Builder sourceDescription = SourceDescription.Builder
                .sourceDescription(typedSource.build());

        // Configuring THEOplayer with defined SourceDescription object.
        theoPlayer.setSource(sourceDescription.build());
        theoPlayer.setAutoplay(true);

        // Adding listeners to THEOplayer basic playback events.
        theoPlayer.addEventListener(PlayerEventTypes.PLAY, event -> Log.i(TAG, "Event: PLAY"));
        theoPlayer.addEventListener(PlayerEventTypes.PLAYING, event -> Log.i(TAG, "Event: PLAYING"));
        theoPlayer.addEventListener(PlayerEventTypes.PAUSE, event -> Log.i(TAG, "Event: PAUSE"));
        theoPlayer.addEventListener(PlayerEventTypes.ENDED, event -> Log.i(TAG, "Event: ENDED"));
        theoPlayer.addEventListener(PlayerEventTypes.ERROR, event -> Log.i(TAG, "Event: ERROR, error=" + event.getError()));
    }

    private void appendMetadata(String metadata) {
        if (viewBinding.metadataTextView.length() == 0) {
            viewBinding.metadataTextView.setText(metadata);
        } else {
            viewBinding.metadataTextView.append("\n\n" + metadata);
        }

        // If metadata content was scrolled to bottom then scroll down automatically after adding new content.
        int fullScrollHeight = viewBinding.metadataScrollView.getScrollY() + viewBinding.metadataScrollView.getHeight();
        int metadataHeight = viewBinding.metadataTextView.getHeight();
        if (metadataHeight <= fullScrollHeight) {
            viewBinding.metadataScrollView.post(() -> viewBinding.metadataScrollView.fullScroll(View.FOCUS_DOWN));
        }
    }


    // In order to work properly and in sync with the activity lifecycle changes (e.g. device
    // is rotated, new activity is started or app is moved to background) we need to call
    // the "onResume", "onPause" and "onDestroy" methods of the THEOplayerView when the matching
    // activity methods are called.

    @Override
    protected void onPause() {
        super.onPause();
        viewBinding.theoPlayerView.onPause();
    }

    @Override
    protected void onResume() {
        super.onResume();
        viewBinding.theoPlayerView.onResume();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        viewBinding.theoPlayerView.onDestroy();
    }

    @Override
    public void onConfigurationChanged(@NonNull Configuration newConfig) {
        super.onConfigurationChanged(newConfig);

        ConstraintSet constraintSet = new ConstraintSet();
        constraintSet.clone(viewBinding.mainConstraintLayout);

        int headerTextViewId = viewBinding.headerTextView.getId();
        int theoPlayerViewId = viewBinding.theoPlayerView.getId();
        int metadataLabelTextViewId = viewBinding.metadataLabelTextView.getId();

        if (newConfig.orientation == Configuration.ORIENTATION_LANDSCAPE) {
            // In landscape orientation metadataLabelTextView (with attached metadataTextView)
            // is placed on the right of headerTextView (with attached theoPlayerView).
            // Additionally metadataLabelTextView width is shrunk to 40% of screen width and its
            // vertical position is centered.
            constraintSet.constrainPercentWidth(headerTextViewId, 0.4F);
            constraintSet.setVerticalBias(headerTextViewId, 0.5F);
            constraintSet.connect(metadataLabelTextViewId, ConstraintSet.START, headerTextViewId, ConstraintSet.END);
            constraintSet.connect(metadataLabelTextViewId, ConstraintSet.TOP, ConstraintSet.PARENT_ID, ConstraintSet.TOP);
        } else {
            // In portrait orientation metadataLabelTextView (with attached metadataTextView)
            // is placed under the theoPlayerView.
            // Additionally metadataLabelTextView width is expanded to full screen width and its
            // vertical alignment is set to top.
            constraintSet.constrainPercentWidth(headerTextViewId, 1F);
            constraintSet.setVerticalBias(headerTextViewId, 0F);
            constraintSet.connect(metadataLabelTextViewId, ConstraintSet.START, ConstraintSet.PARENT_ID, ConstraintSet.START);
            constraintSet.connect(metadataLabelTextViewId, ConstraintSet.TOP, theoPlayerViewId, ConstraintSet.BOTTOM);
        }

        constraintSet.applyTo(viewBinding.mainConstraintLayout);

        // After orientation change metadata content scroll position is adjusted so it could be
        // scrolled automatically when new content is appended.
        viewBinding.metadataScrollView.post(() -> viewBinding.metadataScrollView.fullScroll(View.FOCUS_DOWN));
    }
}

File: metadata-handling/src/main/java/com/theoplayer/sample/playback/metadata/PlayerActivity.kt
package com.theoplayer.sample.playback.metadata

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material3.Card
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.MenuAnchorType
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import com.theoplayer.android.api.THEOplayerConfig
import com.theoplayer.android.api.THEOplayerGlobal
import com.theoplayer.android.api.THEOplayerView
import com.theoplayer.android.api.event.player.ErrorEvent
import com.theoplayer.android.api.event.player.PlayerEventTypes
import com.theoplayer.android.api.event.player.TimeUpdateEvent
import com.theoplayer.android.api.event.track.texttrack.EnterCueEvent
import com.theoplayer.android.api.event.track.texttrack.TextTrackEventTypes
import com.theoplayer.android.api.event.track.texttrack.list.AddTrackEvent
import com.theoplayer.android.api.event.track.texttrack.list.TextTrackListEventTypes
import com.theoplayer.android.api.source.SourceDescription
import com.theoplayer.android.ui.DefaultUI
import com.theoplayer.android.ui.rememberPlayer
import com.theoplayer.android.ui.theme.THEOplayerTheme
import com.theoplayer.sample.common.AppTopBar
import com.theoplayer.sample.common.SourceManager

private data class MetadataOption(
    val nameResId: Int,
    val source: SourceDescription
)

private val METADATA_OPTIONS = listOf(
    MetadataOption(R.string.hlsRadioWithID3MetadataName, SourceManager.HLS_RADIO_WITH_ID3_METADATA),
    MetadataOption(R.string.hlsWithProgramDateTimeMetadataName, SourceManager.HLS_WITH_PROGRAM_DATE_TIME),
    MetadataOption(R.string.hlsWithDateRangeMetadataName, SourceManager.HLS_WITH_DATE_RANGE),
    MetadataOption(R.string.dashWithEmsgMetadataName, SourceManager.DASH_WITH_EMSG),
    MetadataOption(R.string.dashWithScte35MetadataName, SourceManager.DASH_WITH_SCTE35),
)

class PlayerActivity : ComponentActivity() {

    @OptIn(ExperimentalMaterial3Api::class)
    override fun onCreate(savedInstanceState: Bundle?) {
        // Enable all debug logs from THEOplayer.
        THEOplayerGlobal.getSharedInstance(this).logger.enableAllTags()

        super.onCreate(savedInstanceState)

        val initialMetadataId = intent.getIntExtra(PLAYER_PARAM__METADATA_ID, 0)
        val initialIndex = METADATA_OPTIONS.indexOfFirst { it.nameResId == initialMetadataId }.coerceAtLeast(0)

        setContent {
            val context = LocalContext.current
            val metadataEntries = remember { mutableStateListOf<String>() }
            var selectedIndex by remember { mutableIntStateOf(initialIndex) }
            var dropdownExpanded by remember { mutableStateOf(false) }
            val selectedOption = METADATA_OPTIONS[selectedIndex]

            val theoplayerView = remember(context) {
                THEOplayerView(context, THEOplayerConfig.Builder().build()).apply {
                    keepScreenOn = true
                }
            }

            val player = rememberPlayer(theoplayerView)
            val theoPlayer = theoplayerView.player

            LaunchedEffect(selectedIndex) {
                metadataEntries.clear()
                theoPlayer.source = selectedOption.source
                theoPlayer.isAutoplay = true
            }

            // Attach metadata and player listeners once.
            LaunchedEffect(player) {
                var lastCueStartTime = Double.NaN

                // Text track listener for metadata.
                theoPlayer.textTracks.addEventListener(TextTrackListEventTypes.ADDTRACK) { event: AddTrackEvent ->
                    val track = event.track
                    Log.i(TAG, "Event: ADDTRACK, type=${track.type}, label=${track.label}")

                    track.addEventListener(TextTrackEventTypes.ENTERCUE) { cueEvent: EnterCueEvent ->
                        val cueContent = cueEvent.cue.content ?: return@addEventListener
                        val startTime = cueEvent.cue.startTime
                        val text: String? = try {
                            cueContent.getJSONObject("content").getString("text")
                        } catch (_: Exception) {
                            try { cueContent.toString() } catch (_: Exception) { null }
                        }
                        if (!text.isNullOrEmpty()) {
                            Log.i(TAG, "Event: ENTERCUE, type=${track.type}, content=$text")
                            runOnUiThread {
                                // Group cues with the same startTime into one entry.
                                val lastIndex = metadataEntries.lastIndex
                                if (lastIndex >= 0 && lastCueStartTime == startTime) {
                                    metadataEntries[lastIndex] = metadataEntries[lastIndex] + "\n" + text
                                } else {
                                    metadataEntries.add(text)
                                }
                                lastCueStartTime = startTime
                            }
                        }
                    }
                }

                // Listening to 'timeupdate' events that are triggered every time EXT-X-PROGRAM-DATE-TIME
                // is updated.
                theoPlayer.addEventListener(PlayerEventTypes.TIMEUPDATE) { event: TimeUpdateEvent ->
                    if (METADATA_OPTIONS[selectedIndex].nameResId != R.string.hlsWithProgramDateTimeMetadataName) return@addEventListener
                    Log.i(TAG, "Event: TIMEUPDATE, currentTime=${event.currentTime}, programDateTime=${theoPlayer.currentProgramDateTime}")
                    val dateTime = theoPlayer.currentProgramDateTime?.toString()
                    if (!dateTime.isNullOrEmpty()) {
                        runOnUiThread { metadataEntries.add(dateTime) }
                    }
                }

                // Attach player event listeners.
                theoPlayer.addEventListener(PlayerEventTypes.SOURCECHANGE) {
                    Log.i(TAG, "Event: SOURCECHANGE")
                }
                theoPlayer.addEventListener(PlayerEventTypes.CURRENTSOURCECHANGE) {
                    Log.i(TAG, "Event: CURRENTSOURCECHANGE")
                }
                theoPlayer.addEventListener(PlayerEventTypes.LOADEDDATA) {
                    Log.i(TAG, "Event: LOADEDDATA")
                }
                theoPlayer.addEventListener(PlayerEventTypes.LOADEDMETADATA) {
                    Log.i(TAG, "Event: LOADEDMETADATA")
                }
                theoPlayer.addEventListener(PlayerEventTypes.DURATIONCHANGE) {
                    Log.i(TAG, "Event: DURATIONCHANGE")
                }
                theoPlayer.addEventListener(PlayerEventTypes.PLAY) {
                    Log.i(TAG, "Event: PLAY")
                }
                theoPlayer.addEventListener(PlayerEventTypes.PLAYING) {
                    Log.i(TAG, "Event: PLAYING")
                }
                theoPlayer.addEventListener(PlayerEventTypes.PAUSE) {
                    Log.i(TAG, "Event: PAUSE")
                }
                theoPlayer.addEventListener(PlayerEventTypes.SEEKING) {
                    Log.i(TAG, "Event: SEEKING")
                }
                theoPlayer.addEventListener(PlayerEventTypes.SEEKED) {
                    Log.i(TAG, "Event: SEEKED")
                }
                theoPlayer.addEventListener(PlayerEventTypes.WAITING) {
                    Log.i(TAG, "Event: WAITING")
                }
                theoPlayer.addEventListener(PlayerEventTypes.READYSTATECHANGE) {
                    Log.i(TAG, "Event: READYSTATECHANGE")
                }
                theoPlayer.addEventListener(PlayerEventTypes.VOLUMECHANGE) {
                    Log.i(TAG, "Event: VOLUMECHANGE")
                }
                theoPlayer.addEventListener(PlayerEventTypes.ENDED) {
                    Log.i(TAG, "Event: ENDED")
                }
                theoPlayer.addEventListener(PlayerEventTypes.ERROR) { event: ErrorEvent ->
                    Log.i(TAG, "Event: ERROR, error=" + event.errorObject)
                }
            }

            THEOplayerTheme(useDarkTheme = true) {
                Scaffold(
                    topBar = { AppTopBar() }
                ) { padding ->
                    Column(
                        modifier = Modifier
                            .padding(padding)
                            .fillMaxSize()
                    ) {
                        ExposedDropdownMenuBox(
                            expanded = dropdownExpanded,
                            onExpandedChange = { dropdownExpanded = it },
                            modifier = Modifier
                                .fillMaxWidth()
                                .padding(12.dp)
                        ) {
                            TextField(
                                value = stringResource(selectedOption.nameResId),
                                onValueChange = {},
                                readOnly = true,
                                label = { Text(stringResource(R.string.metadataNameLabel)) },
                                trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = dropdownExpanded) },
                                colors = ExposedDropdownMenuDefaults.textFieldColors(),
                                modifier = Modifier
                                    .menuAnchor(MenuAnchorType.PrimaryNotEditable)
                                    .fillMaxWidth()
                            )
                            ExposedDropdownMenu(
                                expanded = dropdownExpanded,
                                onDismissRequest = { dropdownExpanded = false }
                            ) {
                                METADATA_OPTIONS.forEachIndexed { index, option ->
                                    DropdownMenuItem(
                                        text = { Text(stringResource(option.nameResId)) },
                                        onClick = {
                                            selectedIndex = index
                                            dropdownExpanded = false
                                        },
                                        contentPadding = ExposedDropdownMenuDefaults.ItemContentPadding
                                    )
                                }
                            }
                        }

                        // Player
                        DefaultUI(
                            modifier = Modifier
                                .fillMaxWidth()
                                .aspectRatio(16f / 9f),
                            player = player
                        )

                        // Metadata label
                        Text(
                            text = stringResource(R.string.metadataContentLabel),
                            style = MaterialTheme.typography.labelLarge,
                            color = MaterialTheme.colorScheme.onBackground,
                            modifier = Modifier.padding(12.dp)
                        )

                        // Metadata content
                        Card(
                            modifier = Modifier
                                .fillMaxWidth()
                                .weight(1f)
                                .padding(horizontal = 12.dp, vertical = 6.dp)
                        ) {
                            val listState = rememberLazyListState()

                            LaunchedEffect(metadataEntries.size) {
                                if (metadataEntries.isNotEmpty()) {
                                    listState.animateScrollToItem(metadataEntries.lastIndex)
                                }
                            }

                            LazyColumn(
                                state = listState,
                                modifier = Modifier.fillMaxSize()
                            ) {
                                itemsIndexed(metadataEntries) { index, entry ->
                                    if (index > 0) {
                                        HorizontalDivider()
                                    }
                                    Text(
                                        text = entry,
                                        style = MaterialTheme.typography.bodySmall,
                                        modifier = Modifier.padding(8.dp)
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    companion object {
        private val TAG = PlayerActivity::class.java.simpleName
        private const val PLAYER_PARAM__METADATA_ID = "METADATA_ID"

        @JvmStatic
        fun play(context: Context, metadataId: Int) {
            val playIntent = Intent(context, PlayerActivity::class.java)
            playIntent.putExtra(PLAYER_PARAM__METADATA_ID, metadataId)
            context.startActivity(playIntent)
        }
    }
}
